crypto/evp/evp_pkey.c

```diff
@@ -82,7 +89,21 @@ EVP_PKEY *EVP_PKCS82PKEY_ex(const PKCS8_PRIV_KEY_INFO *p8, OSSL_LIB_CTX *libctx,
     len = encoded_len;
     selection = EVP_PKEY_KEYPAIR | EVP_PKEY_KEY_PARAMETERS;
     dctx = OSSL_DECODER_CTX_new_for_pkey(&pkey, "DER", "PrivateKeyInfo",
-                                         NULL, selection, libctx, propq);
+                                         keytype, selection, libctx, propq);
+
+    /* OSSL_DECODER_CTX_get_num_decoders() correctly handles a NULL argument */
```

Comment:
But, `dctx == NULL` indicates an error, and I don't think that it's fruitful to retry the decoder path when that happens.  So the check for the decoder fallback should really be `(dctx != NULL && OSSL_DECODER_CTX_get_num_decoders(dctx) == 0)`